### ISSUE TITLE
JSON.GET path syntax has an issue with filtering using an OR condition when encountering non-existent values

### ISSUE THREAD

**Issue by arunura** [RedisJSON/RedisJSON#614]
name: JSON.GET path syntax has an issue with filtering using an OR condition when encountering non-existent values
description: The issue is specific to OR condition used in the JSONPath during JSON.GET
title: "[Bug]: JSON.GET issue with using OR in JSONPath"
labels: [bug, triage]
body:
  - type: markdown
    attributes:
      value: |
        Thanks for taking the time to fill out this bug report!

  - type: input
    id: operating_system
    attributes:
      label: OS
      description: Operating system and version (eg Ubuntu 20.04, OSX Catalina)
      placeholder: Linux 5.8.0-1041-aws 

  - type: architecutre
    id: cpu_arch
    attributes:
      label: CPU Architecture
      description: CPU Architecture (eg: x86, x86-64, arm, etc).
      placeholder: x86_64

  - type: input
    id: redis_version
    attributes:
      label: Redis Server Version
      description: What version of redis-server are you running (redis-server --version)
      placeholder: 6.2.5

  - type: input
    id: redis_module_version
    attributes:
      label: Redis Module Version
      description: What version of this redis module was installed?
      placeholder: ReJson 20006

  - type: input
    id: was_redis_docker
    attributes:
      label: Docker Installation
      description: What version of this redis module was installed?
      placeholder: Yes

  - type: checkboxes
    id: installmethod
    attributes:
      label: Docker installation
      description: Check this box, if you installed the release via a docker.
      options:
        - label: I installed the dockerhub release

  - type: textarea
    id: whathappened
    label: What happened?
    description: Please tell us what happened, and what you expected to happen.
    placeholder: Tell us what you see!

  - type: textarea
    id: logs
    label: Relevant log output
    description: If release, please copy and paste any relevant log output. This will be automatically formatted into code, so no need for backticks.
    render: shell

The output of the last two executions are incorrect due to the issue. The issue seems to happen when a value (eg: Paris) not present in the JSON object is used in the path query.
```
127.0.0.1:6379> JSON.set testList . "[{\"name\":\"Thomas\",\"city\":\"London\"},{\"name\":\"Arun\",\"city\":\"Glasgow\"},{\"name\":\"Peter\",\"city\":\"Glasgow\"}]"
OK
127.0.0.1:6379> JSON.get testList '$[?(@.city=="London" || @.city=="Glasgow")]'
"[{\"name\":\"Thomas\",\"city\":\"London\"},{\"name\":\"Arun\",\"city\":\"Glasgow\"},{\"name\":\"Peter\",\"city\":\"Glasgow\"}]"
127.0.0.1:6379> JSON.get testList '$[?(@.city=="London" || @.city=="Paris")]'
"[]"
127.0.0.1:6379> JSON.get testList '$[?(@.city=="London" || @.city=="Glasgow" || @.city=="Paris")]'
"[]"
127.0.0.1:6379> 
```

**Comment by TinusChen** [comment]
（1）MULTIPLE path
When the `JSON.get` command accepts multiple path: OK 

```shell
127.0.0.1:6379> JSON.get testList '$[?(@.city=="London")]' '$[?(@.city=="Beijing")]'
{"$[?(@.city==\"London\")]":[{"name":"Thomas","city":"London"}],"$[?(@.city==\"Beijing\")]":[]}
```
（2）SINGLE path
When the `JSON.get` command accepts one path with using an OR condition: NOT-WORK
```shell
127.0.0.1:6379> JSON.get testList '$[?(@.city=="London" || @.city=="Beijing")]'
[]
```

To debugging in `command_json_get()`, you can find some interesting behavior. 

（1）MULTIPLE path
```shell
*-*-* BEGIN: command_json_get()
*-*-* paths parsed: $[?(@.city=="London")]
*-*-* paths parsed: $[?(@.city=="Beijing")]
*-*-* key INPUT: testList
*-*-* value OUTPUT: BulkString("{\"$[?(@.city==\\\"London\\\")]\":[{\"name\":\"Thomas\",\"city\":\"London\"}],\"$[?(@.city==\\\"Beijing\\\")]\":[]}")
*-*-* END: command_json_get()
```

（2）SINGLE path
```shell
*-*-* BEGIN: command_json_get()
*-*-* paths parsed: $[?(@.city=="London" || @.city=="Beijing")]
*-*-* key INPUT: testList
*-*-* value OUTPUT: BulkString("[]")
*-*-* END: command_json_get()
```

Therefore, wrapping operations on multiple results are normal, while some internal mishandling to `single-complicated-JSON-Path-parsing` of these few lines perhaps caused the problem:
https://github.com/RedisJSON/RedisJSON/blob/de32406f63500ccb02dbf392d2062db8862b2ba3/src/commands.rs#L567-L571

When I have more time later I will investigate further or even try to fix.

btw, attached debugging code：

```rust
    println!("*-*-* BEGIN: command_json_get()");
    for path_item in paths.iter() {
        println!("*-*-* paths parsed: {}", path_item);
     }
    
    println!("*-*-* key INPUT: {}", key);
    let key = manager.open_key_read(ctx, &key)?;
    let value = match key.get_value()? {
        Some(doc) => KeyValue::new(doc).to_json(&mut paths, indent, newline, space, format)?,
        None => RedisValue::Null,
    };

    println!("*-*-* value OUTPUT: {:?}", value);
    println!("*-*-* END: command_json_get()");
    Ok(value)
```

**Comment by oshadmi** [comment]
Seems related to a bug in https://github.com/freestrings/jsonpath/issues/87 which is used by RedisJSON

**Comment by abhijit-salvi** [comment]
I am also facing the same issue and was wondering if there's a way around this or a solution yet?

**Comment by oshadmi** [comment]
Will be fixed by #699 (see [test](https://github.com/RedisJSON/RedisJSON/blob/c241b4457103c300b7aac211e76438c1c2b50d8e/tests/rust_tests/filter.rs#L397))
